src/blueprint.zig

```diff
@@ -0,0 +1,281 @@
+		for(self.blocks.items) |block| {
+			new.blocks.appendAssumeCapacity(block);
+		}
+		return new;
+	}
+	pub fn capture(self: *@This(), pos1: Vec3i, pos2: Vec3i) ?struct {x: i32, y: i32, z: i32, message: []const u8} {
+		self.clear();
+
+		const startX = @min(pos1[0], pos2[0]);
+		const startY = @min(pos1[1], pos2[1]);
+		const startZ = @min(pos1[2], pos2[2]);
+
+		const endX = @max(pos1[0], pos2[0]);
+		const endY = @max(pos1[1], pos2[1]);
+		const endZ = @max(pos1[2], pos2[2]);
+
+		const sizeX: usize = @abs(endX - startX + 1);
+		const sizeY: usize = @abs(endY - startY + 1);
+		const sizeZ: usize = @abs(endZ - startZ + 1);
+
+		self.sizeX = sizeX;
+		self.sizeY = sizeY;
+		self.sizeZ = sizeZ;
+
+		self.blocks.ensureCapacity(self.sizeX*self.sizeY*self.sizeZ);
+
+		for(0..sizeX) |offsetX| {
+			const worldX = startX + @as(i32, @intCast(offsetX));
+
+			for(0..sizeY) |offsetY| {
+				const worldY = startY + @as(i32, @intCast(offsetY));
+
+				for(0..sizeZ) |offsetZ| {
+					const worldZ = startZ + @as(i32, @intCast(offsetZ));
+
+					const maybeBlock = main.server.world.?.getBlock(worldX, worldY, worldZ);
+					if(maybeBlock) |block| {
+						self.blocks.appendAssumeCapacity(block);
+					} else {
+						return .{.x = worldX, .y = worldY, .z = worldZ, .message = "Chunk containing block not loaded."};
+					}
+				}
+			}
+		}
+		return null;
+	}
+	pub fn paste(self: @This(), pos: Vec3i) void {
+		const startX = pos[0];
+		const startY = pos[1];
+		const startZ = pos[2];
+
+		var blockIndex: usize = 0;
+
+		for(0..self.sizeX) |offsetX| {
+			const worldX = startX + @as(i32, @intCast(offsetX));
+
+			for(0..self.sizeY) |offsetY| {
+				const worldY = startY + @as(i32, @intCast(offsetY));
+
+				for(0..self.sizeZ) |offsetZ| {
+					const worldZ = startZ + @as(i32, @intCast(offsetZ));
+
+					const block = self.blocks.items[blockIndex];
+					_ = main.server.world.?.updateBlock(worldX, worldY, worldZ, block);
+
+					blockIndex += 1;
+				}
+			}
+		}
+	}
+	pub fn store(self: @This(), allocator: NeverFailingAllocator) []u8 {
+		std.debug.assert(self.sizeX != 0);
+		std.debug.assert(self.sizeY != 0);
+		std.debug.assert(self.sizeZ != 0);
+
+		var gameIdToBlueprintId = self.makeGameIdToBlueprintIdMap(main.stackAllocator);
+		defer gameIdToBlueprintId.deinit();
+		std.debug.assert(gameIdToBlueprintId.count() != 0);
+
+		const blockPalette = self.packBlockPalette(main.stackAllocator, gameIdToBlueprintId);
+		defer main.stackAllocator.free(blockPalette);
+		std.debug.assert(gameIdToBlueprintId.count() == blockPalette.len);
+
+		const blockPaletteSizeBytes = self.getBlockPaletteSize(blockPalette);
+		const blockArraySizeBytes = self.sizeX*self.sizeY*self.sizeZ*@sizeOf(BlockStorageType);
+		const decompressedDataSizeBytes = blockPaletteSizeBytes + blockArraySizeBytes;
+		std.debug.assert(decompressedDataSizeBytes != 0);
+
+		var decompressedWriter = BinaryWriter.initCapacity(allocator, .big, decompressedDataSizeBytes);
+		defer decompressedWriter.deinit();
```

Comment:
`decompressed` implies that it was compressed before, which is not true.
Maybe `uncompressedWriter` or `rawWriter` would be better?